# F21DL Coursework: Part 4 and Part 5

## Preparation

In [1]:
import numpy as np
import pandas as pd
import sklearn
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
np.random.seed(42)
import dill
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

In [ ]:
# # Training data
# with np.load("./kmnist/k49-train-imgs.npz") as xTrain:
#     xTrain = xTrain['arr_0']

# # Training labels
# with np.load("./kmnist/k49-train-labels.npz") as yTrain:
#     yTrain = yTrain["arr_0"]

# # Testing data 
# with np.load("./kmnist/k49-test-imgs.npz") as xTest:
#     xTest = xTest["arr_0"]

# #Testing labels
# with np.load("./kmnist/k49-test-labels.npz") as yTest:
#     yTest = yTest["arr_0"]

In [8]:
# Flatten the 28*28 matrix so that each image is represented in one row.
xTrain = xTrain.reshape(-1, 28*28)
xTest = xTest.reshape(-1, 28*28)

## Linear Classifier

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score 

##### Running Model On Training Data without 10-Fold Cross Validation

In [ ]:
linear = LinearRegression()
linear.fit(xTrain, yTrain);

In [ ]:
linear.intercept_

20.829807562391697

In [ ]:
linear.coef_.shape

(784,)

**Note:** There are 784 coefficients because that is the number of columns/independent variables our data has.

In [ ]:
linear.coef_

array([-4.68610055e-03,  4.79246386e-03, -5.44106119e-03, -2.14820381e-03,
        1.42278488e-03, -1.47018274e-03,  1.76528812e-03,  1.45062913e-03,
        2.73330116e-03,  2.43489326e-03,  3.47632162e-03,  4.12079313e-03,
        3.37185932e-03,  3.85154906e-03,  6.09081295e-03,  4.87109817e-03,
        4.21716934e-03,  4.75402241e-03,  8.83698548e-04,  2.40600081e-03,
        3.63063793e-03, -7.45725525e-04,  1.76858939e-03, -4.70996433e-03,
        4.33262413e-03, -1.13444042e-03, -6.12468594e-03, -1.23485853e-03,
       -3.84606121e-03, -6.95896817e-03,  4.13218419e-03, -3.48936521e-03,
       -1.58430914e-03,  6.51374244e-04, -3.11229725e-03, -2.76855851e-03,
       -5.58352992e-03, -1.21935683e-03, -4.49960411e-03, -3.90674557e-03,
       -4.60952301e-03, -5.24787813e-03, -3.59680155e-03,  3.33328628e-04,
        2.13080387e-03, -7.67054239e-05,  5.96115658e-04,  1.32103066e-03,
       -2.17589767e-03,  5.07680036e-04, -6.25016954e-04, -4.10751419e-04,
       -9.21869779e-04, -

In [ ]:
yPred = linear.predict(xTrain)

In [ ]:
# RMSE
rsme_old=(np.sqrt(mean_squared_error(yTrain, yPred)))

In [ ]:
theta_best_svd.shape

(784,)

In [ ]:
theta_best_svd, residuals, rank, s = np.linalg.lstsq(xTrain, yTrain, rcond=1e-6);

In [ ]:
theta_best_svd # Least-squares solution

array([-8.36818193e-03, -3.80897410e-04, -7.87313568e-03, -2.96385902e-03,
        4.11645699e-04, -9.83842622e-04,  7.67024927e-05,  1.89212169e-03,
        3.60534049e-03,  3.79193957e-03,  5.34597625e-03,  6.61648535e-03,
        5.53960079e-03,  8.10916972e-03,  8.70043841e-03,  8.98558512e-03,
        7.18115539e-03,  7.06620551e-03,  1.70275470e-03,  1.98373383e-03,
        1.53356336e-03, -2.96880343e-03, -1.45974549e-03, -1.02991520e-02,
        1.25426933e-03, -1.41240790e-03, -7.72691151e-03, -1.60794614e-02,
       -1.58922188e-02, -1.56567150e-03,  1.29163399e-03, -6.07609429e-04,
       -2.81860878e-03,  1.96284807e-03, -2.65586824e-03, -2.55817728e-03,
       -5.17916023e-03,  1.80444600e-04, -3.52077163e-03, -3.33356385e-03,
       -4.77505645e-03, -5.17496146e-03, -3.11514110e-03,  3.80121908e-04,
        2.33695195e-03,  1.73577779e-03,  1.51458711e-03,  2.67059255e-03,
        7.32566703e-04,  3.76816211e-03,  3.33513452e-03,  2.66721318e-03,
        5.64002533e-03, -

In [ ]:
R_square_old = r2_score(yTrain, yPred)

In [ ]:
yPred_test = linear.predict(xTest)

In [ ]:
rsme_test_old=(np.sqrt(mean_squared_error(yTest, yPred_test)))

In [ ]:
R_square_test_old = r2_score(yTest, yPred_test)

In [ ]:
print("Evaluation on Training Data")
print(f"RSME: {rsme_old:.3f}")
print(f"R^2 : {R_square:.3f}\n")

print("Evaluation on Test Data")
print(f"RSME: {rsme_test_old:.3f}")
print(f"R^2 : {R_square_test_old:.3f}")

Evaluation on Training Data
RSME: 12.641
R^2 : 0.172

Evaluation on Test Data
RSME: 12.896
R^2 : 0.139


#### Running Model On Training Data with 10-Fold Cross Validation

In [ ]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error

In [ ]:
cv = KFold(n_splits=10, random_state=1, shuffle=True)

In [ ]:
rsme = cross_val_score(linear, xTrain, yTrain, scoring="neg_root_mean_squared_error", cv=cv, n_jobs=-1)

In [ ]:
r_2 = cross_val_score(linear, xTrain, yTrain, scoring="r2", cv=cv, n_jobs=-1)

In [ ]:
rsme_test = cross_val_score(linear, xTest, yTest, scoring="neg_root_mean_squared_error", cv=cv, n_jobs=-1)

In [ ]:
r_2_test = cross_val_score(linear, xTest, yTest, scoring="r2", cv=cv, n_jobs=-1)

In [ ]:
print("Evaluation on Training Data")
print(f"RSME Mean: {-np.mean(rsme):.3f} Standard Deviation: {np.std(rsme):.3f}")
print(f"R^2  Mean: {np.mean(r_2):.3f} Standard Deviation: {np.std(r_2):.3f}\n")

print("Evaluation on Test Data")
print(f"RSME Mean: {-np.mean(rsme_test):.3f} Standard Deviation: {np.std(rsme_test):.3f}")
print(f"R^2  Mean: {np.mean(r_2_test):.3f} Standard Deviation: {np.std(r_2_test):.3f}\n")

Evaluation on Training Data
RSME Mean: 12.687 Standard Deviation: 0.027
R^2  Mean: 0.166 Standard Deviation: 0.003

Evaluation on Test Data
RSME Mean: 12.908 Standard Deviation: 0.122
R^2  Mean: 0.137 Standard Deviation: 0.013



#### Comparison

In [ ]:
print("Evaluation on Training Data")
print("Without Cross Validation                       With Cross Validation")
print(f"RSME: {rsme_old:.3f}                          RSME Mean: {-np.mean(rsme):.3f} Standard Deviation: {np.std(rsme):.3f}")
print(f"R^2 : {R_square_old:.3f}                           R^2  Mean: {np.mean(r_2):.3f} Standard Deviation: {np.std(r_2):.3f}\n") 

print("Evaluation on Test Data")
print("Without Cross Validation                       With Cross Validation")
print(f"RSME: {rsme_test_old:.3f}                         RSME Mean: {-np.mean(rsme_test):.3f} Standard Deviation: {np.std(rsme_test):.3f}")
print(f"R^2 : {R_square_test_old:.3f}                          R^2  Mean: {np.mean(r_2_test):.3f} Standard Deviation: {np.std(r_2_test):.3f}")

Evaluation on Training Data
Without Cross Validation                       With Cross Validation
RSME: 12.641                          RSME Mean: 12.687 Standard Deviation: 0.027
R^2 : 0.172                           R^2  Mean: 0.166 Standard Deviation: 0.003

Evaluation on Test Data
Without Cross Validation                       With Cross Validation
RSME: 12.896                         RSME Mean: 12.908 Standard Deviation: 0.122
R^2 : 0.139                          R^2  Mean: 0.137 Standard Deviation: 0.013


##### Observations
1. Without Cross Validation (column 1 above), RSME **worsens by 0.255** units for the test data. $R^{2}$ also **worsens by 0.033**.
2. With Cross Validation (column 2 above), Mean RSME **worsens by 0.221** units for the test data. Mean $R^{2}$ also **worsens by 0.029**.
3. Although the differences between the training and test data metrics is seemingly small, it is worse nonetheless so it appears that **this linear classifier does not generalize well to new data.**
4. Threshold for good RMSE values depend on the independent variables. I have no idea what might be good here for pixle brightness values, and can't seem to find any online.
5. Assuming this RMSE value is low enough, it would imply a better fitting linear model. However, the really low $R^{2}$ value suggests that the model cannot explain a large proportion of the variance in the data, meaning it is not very good. These contradicting values can occur when we have many degrees of freedom, like 783 in our dataset, which reduces RMSE but just ends up creating a lot of noise in the data so $R^{2}$ goes down.
6. To answer the courseowork spec question, **I would hypothesize ($H_{0}$:) that this dataset is not linearly separable.** 